### ISSUE TITLE
Convert tests using unittest to pytest

### ISSUE THREAD

**Issue by bendnorman** [catalyst-cooperative/pudl#1357]
Our codebase currently uses a mixture of `pytest` and `unittest` to run tests. In order to standardize, let's convert the remaining instances of `unittest` into `pytest`. To do this:

- locate all instances of `unittest`
- write these tests in equivalent `pytest` code. For example, switch from importing MagicMock to calling `mocker.MagicMock`, as we currently do in `test_retry()` in `pudl.test.unit.helpers_test.py`.
- remove imports of `unittest` package

`unittest` shows up in the following places in our codebase:

`$ git grep unittest`

```
test/unit/extract/csv_test.py:from unittest.mock import MagicMock, patch
test/unit/extract/excel_test.py:import unittest
test/unit/extract/excel_test.py:from unittest import mock as mock
test/unit/extract/excel_test.py:class TestMetadata(unittest.TestCase):
test/unit/extract/excel_test.py:class TestExtractor(unittest.TestCase):
test/unit/extract/phmsagas_test.py:from unittest.mock import MagicMock, patch
test/unit/output/ferc1_test.py:import unittest
test/unit/output/ferc1_test.py:class TestForestSetup(unittest.TestCase):
test/unit/transform/classes_test.py:from unittest.mock import MagicMock
test/unit/workspace/datastore_test.py:import unittest
test/unit/workspace/datastore_test.py:class TestDatapackageDescriptor(unittest.TestCase):
test/unit/workspace/datastore_test.py:class TestZenodoFetcher(unittest.TestCase):
test/unit/workspace/resource_cache_test.py:import unittest
test/unit/workspace/resource_cache_test.py:class TestGoogleCloudStorageCache(unittest.TestCase):
test/unit/workspace/resource_cache_test.py:class TestLocalFileCache(unittest.TestCase):
test/unit/workspace/resource_cache_test.py:class TestLayeredCache(unittest.TestCase):
```

**Comment by bendnorman** [comment]
@zaneselvans I'm not going to close this issue because some modules in PUDL are still using unittest instead of pytest. What do you think?

**Comment by zaneselvans** [comment]
Agreed. It would be good to get everything using the same framework.

**Comment by ggurjar333** [comment]
can I TAKE this?

**Comment by e-belfer** [comment]
@zaneselvans Is this still relevant, or was this closed in #3296? I'm not seeing `unittest` in our codebase anywhere. If I've missed something, this needs to be fleshed out to be an actual good first issue.

**Comment by zaneselvans** [comment]
We do still have a number of cases where we're using `unittest`, many of which I think could be refactored to use pytest idioms instead. I guess we should identify which ones we actually want to refactor. This might also be something that CoPilot can do pretty effectively...

```console
$ git grep unittest
```
```
test/unit/extract/csv_test.py:from unittest.mock import MagicMock, patch
test/unit/extract/excel_test.py:import unittest
test/unit/extract/excel_test.py:from unittest import mock as mock
test/unit/extract/excel_test.py:class TestMetadata(unittest.TestCase):
test/unit/extract/excel_test.py:class TestExtractor(unittest.TestCase):
test/unit/extract/phmsagas_test.py:from unittest.mock import MagicMock, patch
test/unit/output/ferc1_test.py:import unittest
test/unit/output/ferc1_test.py:class TestForestSetup(unittest.TestCase):
test/unit/transform/classes_test.py:from unittest.mock import MagicMock
test/unit/workspace/datastore_test.py:import unittest
test/unit/workspace/datastore_test.py:class TestDatapackageDescriptor(unittest.TestCase):
test/unit/workspace/datastore_test.py:class TestZenodoFetcher(unittest.TestCase):
test/unit/workspace/resource_cache_test.py:import unittest
test/unit/workspace/resource_cache_test.py:class TestGoogleCloudStorageCache(unittest.TestCase):
test/unit/workspace/resource_cache_test.py:class TestLocalFileCache(unittest.TestCase):
test/unit/workspace/resource_cache_test.py:class TestLayeredCache(unittest.TestCase):
```

**Comment by e-belfer** [comment]
I updated the issue to clarify what the task actually is!

**Comment by ggurjar333** [comment]
can I TAKE this?

**Comment by e-belfer** [comment]
Hi @ggurjar333, sorry for the slow response! We'll want to switch from importing `MagicMock` to calling `mocker.MagicMock`, as we currently do in `test_retry()` in `pudl.test.unit.helpers_test.py`.

Looks like you've opened a PR - when you resolve the remaining test failures, we'll go ahead and give it a review. Let us know if you get stuck or have any questions! Thanks for taking this on!

**Comment by ggurjar333** [comment]
I'm working on this further to solve errors on classes_test.py

Do we have any more context here? @e-belfer 

<img width="647" alt="{CB4A75EE-742A-436B-8F6B-28A6818BAC18}" src="https://github.com/user-attachments/assets/f7669f34-0336-4fdb-a67e-422268b32b66">

**Comment by zaneselvans** [comment]
@ggurjar333 You can also run `make pytest-unit` which will just run the unit tests without the coverage requirement.

**Comment by ggurjar333** [comment]
@zaneselvans  I think my pytest is not properly getting paths for the root dir or something.

<img width="632" alt="{F26A9FEA-FCAF-430D-A7E7-CC1D101C6125}" src="https://github.com/user-attachments/assets/76a53b46-c23a-456e-8ef2-088458400b9a">

I'm not sure what is happening exactly.
<img width="899" alt="{CDD1EA2B-A694-469E-A4DF-52F174892D13}" src="https://github.com/user-attachments/assets/1a3ae6e9-cfbd-4266-a653-233abfd4f325">

**Comment by zaneselvans** [comment]
What command are you running and from what directory?

**Comment by aurelpere** [comment]
Hi,
I made the reformating in pytest syntax for the two files at the bottom of the list (resource_cache_test.py and datastore_test.py and the tests run well, but my knowledge for networkX is too basic to properly tackle ferc1_test.py
I made the assumption a simple syntax reformating was enough to address the issue as the tests passed when running make pytest-unit, but it seems a bit more tricky.
Regarding the second point of the assignment in the issue:
> * write these tests in equivalent `pytest` code. For example, switch from importing MagicMock to calling `mocker.MagicMock`, as we currently do in `test_retry()` in `pudl.test.unit.helpers_test.py`.
I'm kind of puddled mocker doesnt appear neither in the fixtures in conftest.py nor in the import (for example in classes_test.py)
Also, had a path issue with conftest.py after ferc1_test failed which is resolved moving conftest.py in src directory to force pytest to find pudl module path, but then I have a few errors in the tests:
9 errors in analysis and metadata test files
4 errors for fixtures not found in helpers and settings tests files

I applied to a job offer after reading a bit the code of the repo and thought I would try the test issue as tests usually allow to understand better the code. However here again, i'm puddled so few functions get imported from the tested files! 
I'm not sure the general architecture of the code of this repo fits larger opensource contributing...the technical debt and maintenance must be a lot of work. 
I'm also surprised of so much work done in the tests without even downloading the data files or outputing results in the configured directories...
Will copy this message to answer Dazhong

**Comment by jdangerx** [comment]
Hi Aurel, I'm glad you took a look but @ggurjar333 has already tackled most of the work for this issue in [this PR](https://github.com/catalyst-cooperative/pudl/pull/3931).

The tech debt and maintenance is a lot of work, it's true! We try our best to pay down the tech debt whenever we can, but as I'm sure you know that can be hard to prioritize even in companies that are flush with cash. I have dreams of evolving the architecture towards something where there's a core + an extension system that lets people with lots of dataset-specific knowledge contribute easily. But that's obviously a pretty big project.

As for the actual data files, our current test setup downloads them as part of running the "fast" ETL - the `pudl_io_manager` fixture in `conftest.py` is what coordinates all that.
